community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java

```diff
@@ -228,22 +230,21 @@ private void flushAll()
     /**
      * Insert the given batch of updates into the index defined by the given {@link IndexDescriptor}.
      *
-     * @param descriptor the index descriptor.
+     * @param population the index population.
      * @param batch the list of updates to insert.
      */
-    private void flush( IndexDescriptor descriptor, List<NodePropertyUpdate> batch )
+    private void flush( IndexPopulation population, List<NodePropertyUpdate> batch )
     {
         activeTasks.incrementAndGet();
 
         executor.execute( () -> {
             try
             {
-                IndexPopulation population = populations.get( descriptor );
                 population.populator.add( batch );
             }
-            catch ( Throwable t )
+            catch ( Throwable failure )
             {
-                fail( descriptor, t );
+                populationFailures.add( new PopulationFailure( population.descriptor, failure ) );
             }
             finally
             {
```

Comment:
would it make more sense to `processPopulationFailuresIfAny` in a finally block here instead of on any addApplicable? since it's the only place where we can actually get new failure in a queue of failures? and if it's the case can we avoid queue at all?